### PR TITLE
Unchecked stream subscription and replication.

### PR DESCRIPTION
**Pull request by grixa**
#801, backported to the master.

### CODE REVIEW

File: Blocks/Persistence/file.h [hunk 1]
```diff
@@ -461,6 +461,37 @@ class FilePersister {
...
+    current::locks::SmartMutexLockGuard<MLS> lock(file_persister_impl_->mutex_ref);
+
+    end_t iterator = file_persister_impl_->end.load();
+    const auto tab_pos = entry_json.find('\t');
+    if (tab_pos == std::string::npos) {
+      CURRENT_THROW(MalformedEntryException(entry_json));
+    }
+    const idxts_t idxts = ParseJSON<idxts_t>(entry_json.substr(0, tab_pos));
+    if (idxts.index != iterator.next_index) {
+      CURRENT_THROW(UnsafePublishBadIndexTimestampException(iterator.next_index, idxts.index));
```
Comment (sompylasar) [inline Blocks/Persistence/file.h]: `UnsafePublish` -> `UncheckedPublish` or `PublishUnchecked` ?

File: Blocks/Persistence/memory.h [hunk 2]
```diff
@@ -173,6 +173,28 @@ class MemoryPersister {
...
+    current::locks::SmartMutexLockGuard<MLS> lock(container_->mutex_ref);
+    const auto head = container_->head;
+    const auto tab_pos = entry_json.find('\t');
+    if (tab_pos == std::string::npos) {
+      CURRENT_THROW(MalformedEntryException(entry_json));
+    }
+    const auto idxts = ParseJSON<idxts_t>(entry_json.substr(0, tab_pos));
+    const auto expected_index = static_cast<uint64_t>(container_->entries.size());
+    if (idxts.index != expected_index) {
+      CURRENT_THROW(ss::InconsistentIndexException(expected_index, idxts.index));
```
Comment (sompylasar) [inline Blocks/Persistence/memory.h]: `Unsafe` (/`Unchecked`) `PublishBadIndexTimestampException` here, too?

File: Blocks/SS/persister.h [hunk 3]
```diff
@@ -65,6 +66,15 @@ class EntryPersister : public GenericEntryPersister<ENTRY>, public IMPL {
   IndexAndTimestamp Publish(ENTRY&& e, std::chrono::microseconds us) {
     return IMPL::template DoPublish<MLS>(std::move(e), us);
   }
+  template <current::locks::MutexLockStatus MLS = current::locks::MutexLockStatus::NeedToLock>
+  IndexAndTimestamp PublishUnchecked(const std::string& entry_json) {
```
Comment (sompylasar) [inline Blocks/SS/persister.h]: Add a comment what is unchecked here compared to regular publish?

File: Blocks/SS/persister.h [hunk 4]
```diff
@@ -98,24 +108,28 @@ class EntryPersister : public GenericEntryPersister<ENTRY>, public IMPL {
...
-    return IMPL::template Iterate<IM>(from, std::chrono::microseconds(-1));
+
+  template <current::locks::MutexLockStatus MLS = current::locks::MutexLockStatus::NeedToLock>
+  IterableRange Iterate(std::chrono::microseconds from,
+                        std::chrono::microseconds till = std::chrono::microseconds(-1)) const {
+    return IMPL::template Iterate<MLS>(from, till);
+  }
+
+  template <current::locks::MutexLockStatus MLS = current::locks::MutexLockStatus::NeedToLock>
+  IterableRangeUnsafe IterateUnsafe(std::chrono::microseconds from,
```
Comment (sompylasar) [inline Blocks/SS/persister.h]: Add a comment what is unsafe here compared to regular iterate?

File: Sherlock/pubsub.h [hunk 5]
```diff
@@ -158,6 +158,8 @@ struct ParsedHTTPRequestParams {
   bool entries_only = false;
   // If set, wrap the entries into a large JSON array. Mostly to please JSON-beautifying browser extensions.
   bool array = false;
+  // If set, use "safe" iteration to check the entries before sending.
+  bool checked = false;
```
Comment (sompylasar) [inline Sherlock/pubsub.h]: `If set, use "safe" iteration` -> `If set, use "checked" iteration`?

Comment (sompylasar) [reply Sherlock/pubsub.h]: or rather ```If set to false, use `SubscribeUnchecked` to skip checking the entries before sending.```?

File: Sherlock/sherlock.h [hunk 6]
```diff
@@ -548,16 +620,18 @@ class StreamImpl {
...
-            Subscribe(*http_chunked_subscriber,
-                      begin_idx,
-                      [this, &data, subscription_id]() {
-                        // NOTE: Need to figure out when and where to lock.
-                        // Chat w/ Max about the logic to clean up completed listeners.
-                        // std::lock_guard<std::mutex> lock(inner_data.http_subscriptions_mutex);
-                        data.http_subscriptions[subscription_id].second = nullptr;
-                      });
+        const auto done_callback = [this, &data, subscription_id]() {
+          // NOTE: Need to figure out when and where to lock.
```
Comment (sompylasar) [inline Sherlock/sherlock.h]: `NOTE` -> `TODO(grixa):`?

**Comment by dkorolev** [comment]
To avoid further divergence, I suggest we review this code on its way to `stable` in https://github.com/C5T/Current/pull/801, and then see how to best merge `master` and `stable`.

@grixa @mzhurovich, deal?

**Comment by dkorolev** [comment]
Status update: https://github.com/C5T/Current/pull/801 is long merged, and the merging of `stable` and `master`is now put on hold.

I don't mind having this PR on the list of unmerged for now. @mzhurovich, let's keep it here so far?

**Comment by grixa** [comment]
If we still need these improvements in `master`, I could update this PR with adapted CR-commits from #801

**Comment by dkorolev** [comment]
@mzhurovich @grixa shall we close this PR w/o merging it for now?

**Comment by mzhurovich** [comment]
@dkorolev Yea, let’s close it for now.

**Pull request closed without merging**